go/pkg/client/tree_test.go

```diff
@@ -936,6 +942,74 @@ func TestComputeMerkleTree(t *testing.T) {
 				MaterializeOutsideExecRoot: true,
 			},
 		},
+		{
+			desc: "Intermediate directory relative symlink (preserved)",
+			input: []*inputPath{
+				{path: "foobarDir/foo", fileContents: fooBlob, isExecutable: true},
+				{path: "foobarDir/bar", fileContents: barBlob},
+				{path: "foobarSymDir", isSymlink: true, symlinkTarget: "foobarDir"},
+			},
+			spec: &command.InputSpec{
+				Inputs:              []string{"foobarSymDir", "foobarSymDir/foo", "foobarSymDir/bar"},
```

Comment:
could you add a test where foobarSymDir is not an explicit input? That way you test the logic that adds parent symlinks to the queue.

Reply:
Adding the symlinks to the queue is covered in this test. Including the symlink as an input is required to trigger the bug and ensure it's fixed. If it's not included, it might be converted to a directory with nothing to conflict with.

Reply:
Without looking at the implementation, this test can pass if you're not adding the parent directory as a symlink, because it already exists as an explicit input. This implementation can regress to just skip parent symlinks and add the physical location of the file without this test failing.

Also, not sure I understand your point about not including the symlink. Why would it be converted to a directory if we use preserveSymlinks? Iiuc, we don't add parent symlinks at all today so we should cover that, regardless of whether there is a conflict with another input or not?

Reply:
>Without looking at the implementation, this test can pass if you're not adding the parent directory as a symlink, because it already exists as an explicit input.

If I remove `foobarSymDir` from `input`, the test would error out with `no such file or directory` because it's in `spec.Inputs`.

If I remove `foobarSymDir` from `spec.Input`, it passes with and without this PR. Without it, it's converted into a directory regardless of `Preserved == true` and since the symlink is not an input itself, it's not included. With it, the symlink is preserved regardless if it's an input or an ancestor to one, and the files paths are transformed to be descendants of the symlink target rather than the symlink itself.

I'm not sure if that addresses your concern?

Reply:
So I think it should still be in `input`, just removed from `spec.Input`. I'm surprised the test passes without this PR (or to be clear, this new test in this PR shouldn't pass without the logic in tree.go). Shouldn't `foobarSymDir` be missing from `Symlinks` under `rootDir` without the logic in this PR?

Reply:
Ah, good point. I shorted the call to `evalParentSymlink` and the test failed because `foobarSymDir` appeared as a directory as well as a symlink, but it should have only appeared as a symlink.

Reply:
I added another test where the parent symlink is not an input.